File: andyrb/videoinfo.rb
require 'data_mapper'
require 'json'
require 'pathname'
require 'filesize'
require 'filemagic'
require 'find'
require 'subprocess'
require 'dentaku'
# rubocop:disable Metrics/ModuleLength, Style/CaseIndentation, Lint/UnneededDisable, Style/ConstantName
require_relative 'mood'

module GenerateVideoInfo
  def self.probe(filepath, verbose = false)
    raise 'First argument must be either a string or a pathname' unless filepath.is_a?(String) || filepath.is_a?(Pathname)
    out = nil
    filepath = Pathname.new(filepath) unless filepath.is_a?(Pathname)
    puts Mood.happy("Extracting metadata from #{filepath.basename}") if verbose
    # out = Subprocess.check_output(['ffprobe', '-i', filepath.realpath.to_s, '-hide_banner', '-of', 'json', '-show_streams', '-show_format', '-loglevel', 'quiet']).to_s
    Util::FindApp.which('ffprobe') do |fp|
      raise 'ffprobe not found' unless fp
      raise 'ffprobe found, but is not executable' if fp && !File.executable?(fp)
      out = Subprocess.check_output(%W(#{fp} -i #{filepath.realpath} -hide_banner -of json -show_streams -show_format -loglevel quiet)).to_s
    end
    out = JSON.parse(out)
    yield out if block_given?
    out
  end

  def self.genhash(filename, inputjson, filehash = nil)
    raise 'You must supply either a json string or a pre-parsed hash.' unless inputjson.is_a?(String) || inputjson.is_a?(Hash) || inputjson.is_a?(DataMapper::Collection)

    jsondata = Util.recursive_symbolize_keys(JSON.parse(inputjson)) if inputjson.is_a? String
    jsondata = Util.recursive_symbolize_keys(JSON.parse(inputjson[0][:jsondata])) if inputjson.is_a? DataMapper::Collection
    jsondata = Util.recursive_symbolize_keys(inputjson) if inputjson.is_a? Hash

    filepath = Pathname.new(filename)

    calc = Dentaku::Calculator.new

    fs = lambda do |n|
      out = Filesize.from(n.to_s + 'b').to('Mb').round(2).to_s + 'Mb/s' if n.to_i >= 1_000_000
      out = Filesize.from(n.to_s + 'b').to('Kb').round.to_s + 'Kb/s' if n.to_i.between?(1000, 999_999)
      out = n.to_s + 'b/s' if n.to_i < 1000
      out
    end

    fr = lambda do |n|
      begin
        calc.evaluate(n).to_f.round(2)
      rescue ZeroDivisionError
        nil
      end
    end

    outhash = {}
    outhash[:filename] = filepath.basename.to_s
    outhash[:filehash] = filehash[filepath.realpath.to_s] if filehash
    outhash[:container] = jsondata[:format][:format_name]
    outhash[:duration] = Time.at(jsondata[:format][:duration].to_f).utc.strftime('%H:%M:%S')
    outhash[:duration_raw] = jsondata[:format][:duration]
    outhash[:numstreams] = jsondata[:format][:nb_streams]
    outhash[:bitrate_total] = fs.call(jsondata[:format][:bit_rate])
    outhash[:bitrate_0_raw] =
      case
      when jsondata[:streams][0].key?('bit_rate')
        jsondata[:streams][0]['bit_rate']
      when jsondata[:streams][0][:tags].key?(:BPS)
        jsondata[:streams][0][:tags][:BPS]
      end
    outhash[:bitrate_0] = fs.call(outhash[:bitrate_0_raw])
    outhash[:type_0] = jsondata[:streams][0][:codec_type] if jsondata[:streams][0].key?(:codec_type)
    outhash[:codec_0] = jsondata[:streams][0][:codec_name] if jsondata[:streams][0].key?(:codec_name)
    outhash[:bitrate_1_raw] =
      case
      when jsondata[:streams][1].respond_to?(:key?) && jsondata[:streams][1].key?('bit_rate')
        jsondata[:streams][1]['bit_rate']
      when jsondata[:streams][1].respond_to?(:key?) && jsondata[:streams][1].key?(:tags) && jsondata[:streams][1][:tags].key?(:BPS)
        jsondata[:streams][1][:tags][:BPS]
      end # rubocop:disable Lint/EndAlignment

    outhash[:bitrate_1] = fs.call(outhash[:bitrate_1_raw])
    # outhash[:bitrate_1] = case
    # when jsondata[:streams].length >= 2 && outhash[:bitrate_1_raw].to_i >= 1_000_000
    #   Filesize.from(outhash[:bitrate_1_raw].to_s + 'b').to('Mb').round(2).to_s + 'Mb/s'
    # when jsondata[:streams].length >= 2 && outhash[:bitrate_1_raw].to_i.between?(1000, 999_999)
    #   Filesize.from(outhash[:bitrate_1_raw].to_s + 'b').to('Kb').round.to_s + 'Kb/s'
    # when jsondata[:streams].length >= 2 && outhash[:bitrate_1_raw].to_i < 1000
    #   outhash[:bitrate_1_raw].to_s + 'b/s'
    # end # rubocop:disable Lint/EndAlignment

    outhash[:type_1] = jsondata[:streams][1][:codec_type] if jsondata[:streams].length >= 2 && jsondata[:streams][1].respond_to?(:key?) && jsondata[:streams][1].key?(:codec_type)

    outhash[:codec_1] = jsondata[:streams][1][:codec_name] if jsondata[:streams].length >= 2 && jsondata[:streams][1].respond_to?(:key?) && jsondata[:streams][1].key?(:codec_name)

    outhash[:height] =
      case
      when jsondata[:streams][0][:height]
        jsondata[:streams][0][:height]
      when jsondata[:streams].length >= 2 && jsondata[:streams][1].respond_to?(:key) && jsondata[:streams][1].key?(:height)
        jsondata[:streams][1][:height]
      end

    outhash[:width] =
      case
      when jsondata[:streams][0].key?(:width)
        jsondata[:streams][0][:width]
      when jsondata[:streams].respond_to?(:length) && jsondata[:streams].length >= 2 && jsondata[:streams][1].is_a?(Hash) && jsondata[:streams][1].key?(:width)
        jsondata[:streams][1][:width]
      end

    outhash[:frame_rate] =
      case # Used 4 'when' statements for readability
      when jsondata.respond_to?(:dig) && jsondata.dig(:streams, 0, :avg_frame_rate)
        fr.call(jsondata[:streams][0][:avg_frame_rate])
      when !jsondata.respond_to?(:dig) && jsondata.key?(:streams) && jsondata[:streams][0] && jsondata[:streams][0].key?(:avg_frame_rate)
        fr.call(jsondata[:streams][0][:avg_frame_rate])
      when jsondata.respond_to?(:dig) && jsondata.dig(:streams, 1, :avg_frame_rate)
        fr.call(jsondata[:streams][1][:avg_frame_rate])
      when !jsondata.respond_to?(:dig) && !jsondata[:streams][0].key?(:avg_frame_rate) && jsondata[:streams].length >= 2 && jsondata[:streams][1].key?(:avg_frame_rate)
        fr.call(jsondata[:streams][1][:avg_frame_rate])
      end
    yield outhash if block_given?
    outhash
  end

  def self.genfilelist(filelist, testmode = false, sort = true)
    whitelist = %w(video/x-flv video/mp4 video/mp2t video/3gpp video/quicktime video/x-msvideo video/x-ms-wmv video/webm video/x-matroska video/3gpp2 audio/x-wav)
    whitelist += %w(audio/wave video/dvd video/mpeg application/vnd.rn-realmedia-vbr audio/vnd.rn-realaudio audio/x-realaudio)
    magic = FileMagic.new(:mime_type)
    filelist = Util::SortEntries.sort(filelist) if sort
    puts 'Files to be examined:' if testmode
    outlist = []
    filelist.each do |entry|
      if testmode
        puts magic.flags
        puts magic.file(entry)
      end
      outlist << entry if whitelist.include?(magic.file(entry)) && !testmode
      puts Mood.happy { entry } if whitelist.include?(magic.file(entry)) && testmode
    end
    magic.close
    outlist = Util::SortEntries.sort(outlist) if sort
    yield outlist if block_given?
    outlist
  end

  def self.find(directory, verbose = false)
    magic = FileMagic.new
    dirpath = Pathname.new(directory)
    blacklist = /jpg|gif|png|flv|mp4|mkv|webm|vob|ogg|drc|avi|wmv|yuv|rm|rmvb|asf|m4v|mpg|mp2|mpeg|mpe|mpv|3gp|3g2|mxf|roq|nsv|f4v|wav|ra|mka|pdf|odt|docx|webp|swf|cb7|zip|7z|xml|log/i
    initdirectories = dirpath.find.to_a
    initdirectories.keep_if(&:file?)
    initdirectories.delete_if { |i| i.extname =~ blacklist }
    initdirectories.keep_if do |i|
      result = magic.file(i.to_s)
      result.include?('SQLite 3.x')
    end
    directories = []
    initdirectories.each do |i|
      puts Mood.happy { "Adding #{i.dirname} to list" } if verbose
      directories << i.dirname
    end
    directories.uniq!
    yield directories if block_given?
    directories
  end

  class Videoinfo
    include DataMapper::Resource

    storage_names[:default] = 'videoinfo'

    property :id, Serial
    property :filename, Text, lazy: false, key: true, unique: true
    property :duration, String
    property :duration_raw, Float
    property :numstreams, Integer
    property :container, Text, lazy: false
    property :bitrate_total, String
    property :width, Integer
    property :height, Integer
    property :frame_rate, Float
    property :bitrate_0, String
    property :bitrate_0_raw, Integer
    property :type_0, String
    property :codec_0, String
    property :bitrate_1, String
    property :bitrate_1_raw, Integer
    property :type_1, String
    property :codec_1, String
    property :filehash, String, key: true, unique: true, lazy: false
  end

  class Videojson
    include DataMapper::Resource

    storage_names[:default] = 'videojson'

    property :id, Serial
    property :filename, Text, lazy: false, key: true, unique: true
    property :jsondata, Text, lazy: false
  end

  class Data
    def initialize(dbpath, verbose = false)
      @dbpath = Pathname.new(dbpath)
      @verbose = verbose
      DataMapper.setup(:default, "sqlite:#{@dbpath.realpath}")
      DataMapper::Logger.new($stdout, :debug) if @verbose
      @db = DataMapper.repository(:default).adapter
      @vi = Videoinfo.new
      @vj = Videojson.new
      DataMapper.finalize
      DataMapper::Model.raise_on_save_failure = true
    end

    def existing
      out = nil
      out = Videoinfo.all(fields: [:filename, :filehash]).to_a if Videoinfo.count >= 1
      # print out.inspect
      out
    end

    def json(filepath, repo = nil)
      raise 'First argument must be either a string or a pathname' unless filepath.is_a?(String) || filepath.is_a?(Pathname)
      out = nil
      Videojson.storage_names[repo] = 'videojson' if repo
      filepath = Pathname.new(filepath) unless filepath.is_a?(Pathname)
      insert = Videojson.new
      # puts Videojson.count(filename: filepath.basename.to_s)
      if @db.storage_exists?('videojson') && Videojson.count(filename: filepath.basename.to_s) >= 1
        puts Mood.happy("Reading metadata from cache for #{filepath}") if @verbose
        out = Videojson.all(filename: filepath.basename, fields: [:jsondata])
      else
        puts Mood.happy("Extracting metadata from #{filepath.basename}") if @verbose
        # out = Subprocess.check_output(['ffprobe', '-i', filepath.realpath.to_s, '-hide_banner', '-of', 'json', '-show_streams', '-show_format', '-loglevel', 'quiet']).to_s
        Util::FindApp.which('ffprobe') do |fp|
          raise 'ffprobe not found' unless fp
          raise 'ffprobe found, but is not executable' if fp && !File.executable?(fp)
          out = Subprocess.check_output(%W(#{fp} -i #{filepath.realpath} -hide_banner -of json -show_streams -show_format -loglevel quiet)).to_s
        end
        cache = JSON.parse(out)
        begin
          puts Mood.happy("Caching JSON for #{filepath.basename}") if @verbose
          insert.attributes = { filename: filepath.basename, jsondata: JSON.generate(cache) }
          # print @vi.attributes
          insert.save
          # print "\n"
        rescue DataMapper::SaveFailureError
          insert.errors.each { |e| puts e } if @verbose
          raise "Save failure error raised for #{filepath.basename}" if @verbose
        end
      end
      out
    end
  end
end

File: andy2/videoinfo.py
# pylint: disable=r0903, c0301, c0111, r0912, r0915, c0411
import collections  # noqa: F401  # pylint: disable=W0611
import concurrent.futures  # noqa: F401  # pylint: disable=W0611
import json
import os
import pathlib
import shutil
import time
from contextlib import contextmanager

from humanize.filesize import naturalsize
from sqlalchemy import Column, Float, Integer, String, create_engine  # ForeignKey would go here.
from sqlalchemy.engine.reflection import Inspector
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker  # relationship would go here.
# from sqlalchemy.schema import Table
import sqlalchemy.orm.exc
from andy2.string_eval import NumericStringParser

try:
    from util import Mood, Program, Util  # noqa: F401  # pylint: disable=W0611
except ImportError:
    from andy.util import Mood, Program, Util  # noqa: F401  # pylint: disable=W0611

try:
    # This must be the filemagic module, the python-magic plugin (which uses the same module name) will most likely not work.
    import magic  # noqa: F401  # pylint: disable=W0611
except ImportError:
    pass


@contextmanager
def sqa_session(basesess):
    try:
        yield basesess
        basesess.commit()
    except:
        basesess.rollback()
        raise
    finally:
        basesess.close()


SQLBase = declarative_base()  # pylint: disable=c0103


class VideoInfo(SQLBase):
    __tablename__ = 'videoinfo'
    id = Column(Integer, primary_key=True)  # pylint: disable=c0103
    filename = Column(String, nullable=False, unique=True)
    duration = Column(String)
    duration_raw = Column(Float)
    numstreams = Column(Integer)
    container = Column(String, nullable=False)
    width = Column(Integer)
    height = Column(Integer)
    frame_rate = Column(Float)
    bitrate_total = Column(String)
    bitrate_0 = Column(String)
    bitrate_0_raw = Column(Integer)
    type_0 = Column(String)
    codec_0 = Column(String)
    bitrate_1 = Column(String)
    bitrate_1_raw = Column(Integer)
    type_1 = Column(String)
    codec_1 = Column(String)
    filehash = Column(String, unique=True, nullable=False)

    def __repr__(self):
        return "<VideoInfo(filename={}, duration={}, duration_raw={}, numstreams={}, container={}, width={}, height={}, frame_rate={}, bitrate_total={}, bitrate_0={}, bitrate_0_raw={}, type_0={}, codec_0={}, bitrate_1={}, bitrate_1_raw={}, type_1={}, codec_1={}, filehash={})>".format(self.filename, self.duration, self.duration_raw, self.numstreams, self.container, self.width, self.height, self.frame_rate, self.bitrate_total, self.bitrate_0, self.bitrate_0_raw, self.type_0, self.codec_0, self.bitrate_1, self.bitrate_1_raw, self.type_1, self.codec_1, self.filehash)


class VideoJSON(SQLBase):
    __tablename__ = 'videojson'
    id = Column(Integer, primary_key=True)
    filename = Column(String, nullable=False, unique=True)
    json = Column(String, nullable=False, unique=True)

    def __repr__(self):
        return "<VideoJSON(filename={}, json={})>".format(self.filename, self.json)


class VideoData:
    def __init__(self, db, verbose=False, regen=False, regenjson=False):
        if not pathlib.Path(db).exists():
            pathlib.Path(db).touch()

        self.dbpath = pathlib.Path(db).resolve()
        self.verbose = verbose
        self.dataengine = create_engine("sqlite:///{}".format(self.dbpath))
        self.inspect = Inspector.from_engine(self.dataengine)
        if regen:
            # with self.dataengine.connect() as con:
            #    print("{} Deleting existing videoinfo table.".format(Mood.happy()))
            #    con.execute("DROP TABLE videoinfo")
            #    if regenjson:
            #        print("{} Deleting existing videojson table.".format(Mood.happy()))
            #        con.execute("DROP TABLE videojson")
            #    con.execute('VACUUM')
            print("{} Deleting existing videoinfo table.".format(Mood.happy()))
            VideoInfo.__table__.drop(self.dataengine)  # pylint: disable=e1101
            if regenjson:
                print("{} Deleting existing videojson table.".format(Mood.happy()))
                VideoJSON.__table__.drop(self.dataengine)   # pylint: disable=e1101
        print("{} Creating database tables.".format(Mood.happy()))
        SQLBase.metadata.create_all(self.dataengine)
        if verbose:
            print("{} Tables List:\n{}".format(Mood.happy(), self.inspect.get_table_names()))
            if "videoinfo" in self.inspect.get_table_names():
                print("{} Column Names:\n{}".format(Mood.happy(), self.inspect.get_columns("videoinfo")))
        sessionbase = sessionmaker(bind=self.dataengine)
        self.session = sessionbase()  # pylint: disable=c0103

    def parse(self, videofile, probe=None, quiet=False):
        cache = None
        try:
            cache = self.session.query(VideoJSON).filter(VideoJSON.filename == videofile).one()
            print("{} Information found in the cache.".format(Mood.happy()))
            # print(cache.json)
        except sqlalchemy.orm.exc.NoResultFound:
            # if self.verbose:
            #    print("{} No entry in the cache.".format(Mood.neutral()))
            print("{} No entry in the cache.".format(Mood.neutral()))

        if isinstance(cache, VideoJSON):
            return cache.json
        else:
            ffprobe = None
            if probe:
                ok = os.access(probe, mode=os.X_OK, effective_ids=True)  # pylint: disable=c0103
                if ok:
                    ffprobe = probe
                    del probe
                    del ok
                else:
                    raise FileNotFoundError("Specified ffprobe compatible command not found or not executable by current user.")
            else:
                ffprobe = shutil.which("ffprobe", mode=os.X_OK)

            if not ffprobe:
                raise FileNotFoundError("Could not find ffprobe.")
            if not quiet:
                print("{} Extracting information from {}".format(Mood.happy(), videofile))
            return json.loads(Program.returninfo([ffprobe, "-i", videofile, "-hide_banner", "-of", "json", "-show_streams", "-show_format"], string=True))

    @staticmethod
    def probe(videofile, prog=None, quiet=False):
        """Alternate version of the parse method that does not use the database.

        videofile is the file to probe.

        prog is the program to use to do the probing. If none is specified, ffprobe shall be searched for and used."""

        ffprobe = None
        if prog:
            ok = os.access(prog, mode=os.X_OK, effective_ids=True)  # pylint: disable=c0103
            if ok:
                ffprobe = prog
                del prog
                del ok
            else:
                raise FileNotFoundError("Specified ffprobe compatible command not found or not executable by current user.")
        else:
            ffprobe = shutil.which("ffprobe", mode=os.X_OK)

        if not ffprobe:
            raise FileNotFoundError("Could not find ffprobe.")
        if not quiet:
            print("{} Extracting information from {}".format(Mood.happy(), videofile))
        return json.loads(Program.returninfo([ffprobe, "-i", videofile, "-hide_banner", "-of", "json", "-show_streams", "-show_format"], string=True))

    @classmethod
    def cwd(cls, verbosemode=False, regen=False, regenjson=False):  # pylint: disable=W0221
        """Class method to simplify and prettify accessing a videoinfo database in the current directory with the name "videoinfo.sqlite".

        verbosemode takes a True or False and passes it along to the parent class."""
        if not pathlib.Path.cwd().joinpath("videoinfo.sqlite").exists():  # pylint: disable=e1101
            pathlib.Path.cwd().joinpath("videoinfo.sqlite").touch()  # pylint: disable=e1101
        return cls(str(pathlib.Path.cwd().joinpath("videoinfo.sqlite")), verbosemode, regen, regenjson)

    @staticmethod
    def gendict(filename, jsondata, filehash):
        if not isinstance(jsondata, dict):
            jsondata = json.loads(jsondata)

        yield "filename", pathlib.Path(filename).name
        yield "hash", filehash
        yield "container", jsondata["format"]["format_name"]
        yield "duration", time.strftime("%H:%M:%S", time.gmtime(int(float(jsondata["format"]["duration"]))))
        yield "duration_raw", jsondata["format"]["duration"]
        yield "numstreams", int(jsondata["format"]["nb_streams"])
        yield "codec_0", jsondata["streams"][0]["codec_name"]
        yield "type_0", jsondata["streams"][0]["codec_type"]
        if isinstance(jsondata["streams"][1], dict) and jsondata["streams"][1]["codec_name"]:
            yield "codec_1", jsondata["streams"][1]["codec_name"]
        else:
            yield "codec_1", None

        if isinstance(jsondata["streams"][1], dict) and jsondata["streams"][1]["codec_type"]:
            yield "type_1", jsondata["streams"][1]["codec_type"]
        else:
            yield "type_1", None

        def bitrate(stream):
            if not isinstance(stream, int):
                raise TypeError("Argument must be an integer.")
            try:
                if isinstance(jsondata["streams"][stream], dict):
                    if "tags" in jsondata["streams"][stream] and "bit_rate" not in jsondata["streams"][stream] and "BPS" in jsondata["streams"][stream]["tags"]:
                        return naturalsize(jsondata["streams"][stream]["tags"]["BPS"]).replace(" MB", "Mbps").replace(" kB", "Kbps")
                    elif "bit_rate" in jsondata["streams"][stream]:
                        return naturalsize(jsondata["streams"][stream]["bit_rate"]).replace(" MB", "Mbps").replace(" kB", "Kbps")
                    else:
                        return None
            except (KeyError, IndexError):
                return None
        yield "bitrate_0", bitrate(0)
        yield "bitrate_1", bitrate(1)
        yield "bitrate_total", naturalsize(jsondata["format"]["bit_rate"]).replace(" MB", "Mbps").replace(" kB", "Kbps")

        def bitrate_raw(stream):
            if not isinstance(stream, int):
                raise TypeError("Argument must be an integer.")

            try:
                if isinstance(jsondata["streams"][stream], dict):
                    if "tags" in jsondata["streams"][stream] and "bit_rate" not in jsondata["streams"][stream] and "BPS" in jsondata["streams"][stream]["tags"]:
                        return int(jsondata["streams"][stream]["tags"]["BPS"])
                    elif "bit_rate" in jsondata["streams"][stream]:
                        return int(jsondata["streams"][stream]["bit_rate"])
                    else:
                        return None
            except (KeyError, IndexError):
                return None

        yield "bitrate_0_raw", bitrate_raw(0)
        yield "bitrate_1_raw", bitrate_raw(1)

        def height():
            try:
                if jsondata["streams"][0]["codec_type"] == "video" and jsondata["streams"][0]["height"]:
                    return jsondata["streams"][0]["height"]
                elif isinstance(jsondata["streams"][1], dict) and jsondata["streams"][1]["codec_type"] == "video" and jsondata["streams"][0]["height"]:
                    return jsondata["streams"][1]["height"]
                else:
                    return None
            except (KeyError, IndexError):
                return None

        def width():
            try:
                if jsondata["streams"][0]["codec_type"] == "video" and jsondata["streams"][0]["width"]:
                    return jsondata["streams"][0]["width"]
                elif isinstance(jsondata["streams"][1], dict) and jsondata["streams"][1]["codec_type"] == "video" and jsondata["streams"][1]["width"]:
                    return jsondata["streams"][1]["width"]
            except (KeyError, IndexError):
                return None

        def frame_rate():
            try:
                nsp = NumericStringParser()
                if jsondata["streams"][0]["codec_type"] == "video" and jsondata["streams"][0]["avg_frame_rate"]:
                    return "{0:.2f}".format(float(nsp.eval(jsondata["streams"][0]["avg_frame_rate"])))
                elif isinstance(jsondata["streams"][1], dict) and jsondata["streams"][1]["codec_type"] == "video" and jsondata["streams"][1]["avg_frame_rate"]:
                    return "{0:.2f}".format(float(nsp.eval(jsondata["streams"][1]["avg_frame_rate"])))
                else:
                    return None
            except (KeyError, IndexError):
                return None
        yield "height", height()
        yield "width", width()
        yield "frame_rate", frame_rate()
        yield "jsondata", json.dumps(jsondata)

    def genexisting(self):
        """Generator function that queries an existing videoinfo database and yields the filename and hash for any existing files in the database."""
        if self.session.query(VideoInfo).count() >= 1:
            out = self.session.query(VideoInfo).all()
            # print(out)
            for i in out:
                # print(i.filename)
                # print(i.filehash)
                yield i.filename, i.filehash
        else:
            return None

    def genfilelist(self, filelist, existinghash=None):
        """Generator function that takes a list of files and yields a filtered list that eliminates any non-video files (based on known mime types or file extensions) and any files that are already in the database.
        It will use the filemagic module if available for matching based on mime type or use a file extension whitelist if filemagic is not detected.
        python-magic WILL NOT WORK and there is no easy way to test for it as it uses the same module name.
        So if python-magic is installed, get rid of it and install filemagic instead."""

        try:
            whitelist = ['video/x-flv', 'video/mp4', 'video/mp2t', 'video/3gpp', 'video/quicktime', 'video/x-msvideo', 'video/x-ms-wmv', 'video/webm', 'video/x-matroska', 'video/msvideo', 'video/avi', 'application/vnd.rm-realmedia', 'audio/x-pn-realaudio', 'audio/x-matroska', 'audio/ogg', 'video/ogg', 'audio/vorbis', 'video/theora', 'video/3gpp2', 'audio/x-wav', 'audio/wave', 'video/dvd', 'video/mpeg', 'application/vnd.rn-realmedia-vbr', 'audio/vnd.rn-realaudio', 'audio/x-realaudio']

            with magic.Magic(flags=magic.MAGIC_MIME_TYPE) as m:
                for filename in filelist:
                    filepath = pathlib.Path(filename)
                    if not self.verbose and existinghash:
                        if m.id_filename(filename) in whitelist and filepath.is_file() and filepath.name not in existinghash:
                            yield str(filepath)
                    elif self.verbose or not existinghash:
                        if m.id_filename(filename) in whitelist and filepath.is_file():
                            yield str(filepath)
        except NameError:
            whitelist = ['.webm', '.mkv', '.flv', '.vob', '.ogg', '.drc', '.avi', '.wmv', '.yuv', '.rm', '.rmvb', '.asf', '.mp4', '.m4v', '.mpg', '.mp2', '.mpeg', '.mpe', '.mpv', '.3gp', '.3g2', '.mxf', '.roq', '.nsv', '.f4v', '.wav', '.ra', '.mka']
            for filename in filelist:
                filepath = pathlib.Path(filename)
                if not self.verbose and existinghash:
                    if filepath.suffix in whitelist and filepath.is_file() and filepath.name not in existinghash:
                        yield str(filepath)
                elif self.verbose or not existinghash:
                    if filepath.suffix in whitelist and filepath.is_file():
                        yield str(filepath)

    @staticmethod
    def genhashlist(files, existinghash=None):
        """Generator function that takes a list of files and a list of existing hashes (if any) and calculates hashes for those files.

        files takes a list containing file names for which hashes will be calculated.

        existinghash takes a dictionary where the filename is the key and the hash is the value, this is optional."""

        for filename in files:
            if existinghash and filename not in existinghash:
                print("{} Calculating hash for {}".format(Mood.happy(), pathlib.Path(filename).name))
                yield filename, Util.hashfile(filename)
            else:
                print("{} Calculating hash for {}".format(Mood.happy(), pathlib.Path(filename).name))
                yield filename, Util.hashfile(filename)


class FindVideoInfo:  # pylint: disable=R0903

    """This class contains any functions related to locating videoinfo databases.
    Requires filemagic, python-magic will not work, if python-magic is installed, get rid of it and use filemagic instead.
    It's not easy to test for which is which as they use the same module name."""

    def __init__(self):
        try:
            test = magic.Magic()
        except NameError:
            print("{} Filemagic module not installed.".format(Mood.sad()))
            raise
        else:
            del test

    @staticmethod
    def find(directory="/data/Private"):
        """Worker function that locates directories with videoinfo database that are under the specified directory.
        Files will be run through filemagic to verify that they actually sqlite databases.
        Just like the genfilelist function in GenVideoInfo, this only works with filemagic,
        if python-magic is installed, get rid of it and use filemagic instead. Its not easy to distinguish
        python-magic from filemagic as they use the same module name.

        directory is the directory to be searched."""

        paths = pathlib.Path(directory).rglob("videoinfo.sqlite")
        for filename in paths:
            with magic.Magic() as m, open(str(filename), "rb") as f:
                filetype = m.id_buffer(f.read())
                if "SQLite 3.x database" in filetype:
                    yield str(filename.parent)
